### ISSUE TITLE
Elevation data is aberrant ?

### ISSUE THREAD

**Issue by ABelliqueux** [BourgeoisLab/GPXLab#11]
After getting elevation data from database, the values seem quite aberrant. For instance at some points, I get a high elevation on the beach ( the location is situated in France ).
I get similarly off set values on various gpx files ( sometimes higher, sometimes lower than the "real" values.)

For example, a point with the coordinates 47.591250, -2.721889 ( 47°35'28.5"N 2°43'18.8"W ) gets a 32 meters elevation in GPXlab . Checking the same location in https://www.calcmaps.com/map-elevation/ gives an elevation of 1 meter, which is much closer to reality.

Am I wrong in assuming the elevation data is relative to sea level (ie : sea level = 0 m) ?

**Comment by BourgeoisLab** [comment]
Hi

The elevation data is taken from NASA SRTM (https://www2.jpl.nasa.gov/srtm/). Outside the US there is a datapoint every 30m. In between the program interpolates the data. So if you are near a cliff for example, you may get a higher value. Maybe this is the probelm here.

I will probably not work on this issue, but I would support anyone wanting to work on it. Maybe the program could use google data (more precise), although I am not really fan of this...

**Comment by ABelliqueux** [comment]
Hi ! Thanks for your answer.

I've been able to look for more informations on the subject. It seems that the NASA SRTM should be available in SRTM1 (indeed 30 Meters samples ) for Europe and some other regions.

Yet it seems that in srtm.cpp, we're downloading SRTM3 data (90 Meters sample):
https://github.com/BourgeoisLab/GPXLab/blob/28b0b3014d2728e866ce1ffad21b42acd8c3680b/GPXLab/functions/srtm.cpp#L25

After looking for the SRTM1 data for Europe on this server, I could not find it, yet it should exist, as I found at least one site using it, that generates a more plausible elevation profile for the same file I've been trying in GPXLab ; https://www.gpsvisualizer.com/profile_input

You can see in the dropdown menu "Add DEM elevation data:" that you have a choice between SRTM1 for Europe and North America, and SRTM3 for Worldwide.

The author specifically say on https://www.gpsvisualizer.com/elevation

> NASA's SRTM1 database (from the Space Shuttle Radar Topography Mission) has a resolution of about 30 meters. Previously, SRTM1 data was only available for the United States (and was not very useful because NED1 is generally better), but recently NASA has released SRTM1 tiles for the entire world. GPS Visualizer's server does not contain the full global dataset, but it does contain all of Europe, Japan, South Korea, and Taiwan, as well as parts of Australia, New Zealand, and South Africa. Additional locations will be installed upon request.

I wonder where this SRTM1 data for Europe is stored... 

EDIT : Looks like it's available here, but needs an account ? 

https://e4ftl01.cr.usgs.gov/MEASURES/SRTMGL1.003

More infos : https://earthdata.nasa.gov/nasa-shuttle-radar-topography-mission-srtm-version-3-0-global-1-arc-second-data-released-over-asia-and-australia

I tried to set `SRTMModel model = OneArcSecond` in srtm.h and used the hgt (25,9 MB vs 2,9MB) files from above link after creating an account, but no change in the elevation profile.  Is there something else I have to set in order to use 1 arc/second hgt file ?

Also, Sorry for spamming this thread with edits, I'm not at all fluent in C++ and just trying some things as I have some time to spare !

**Comment by BourgeoisLab** [comment]
Thanks for the info. You are right, I use strm3. Long time since I implemented it... I think I used it because it was available for the whole world (at least at that time).

Strm1 (strm30?) files seems to be available here https://dds.cr.usgs.gov/srtm/version2_1/SRTM30/

I will check this when I have some time. Not sure if it covers the whole world, or if I have to mix between both series. If you can gather more you're welcome to do so.

**Comment by ABelliqueux** [comment]
Hi, 

According to this document ; https://dds.cr.usgs.gov/srtm/version2_1/Documentation/SRTM_Topo.pdf , section 2.2 :

> Sample spacing for individual data points is either 1 arc-second, 3 arc-seconds, or 30 arc-seconds, referred to as SRTM1, SRTM3 and SRTM30, respectively. 

I think SRTMGL1 stands for SRTM GLobal 1arc/s .

I downloaded such a file (attached fro study), wich once unzipped has a size of 25,9 Mo (25 934 402 octets) , which is consistant with the description of a 1 arc/s file you can find in srtm.h, as copied from https://www2.jpl.nasa.gov/srtm/faq.html : 

> International 3-arc-second files have 1201 columns and 1201 rows of data, with a total file size of 2,884,802 bytes ( = 1201 x 1201 x 2). United States 1-arc-second files have 3601 columns and 3601 rows of data, with a total file size of 25,934,402 bytes ( = 3601 x 3601 x 2).

So I think SRTMGL1 hgt files are really 1a/s files for worldwide.

I attach a SRTGL1 HGT file for study if you want to have a look at actual data :

[N47W003.SRTMGL1.hgt.zip](https://github.com/BourgeoisLab/GPXLab/files/3597164/N47W003.SRTMGL1.hgt.zip)

**Comment by BourgeoisLab** [comment]
In a quick test I saw that you can use the file you provided and change in dialog_srtm.cpp line 29 to
`srtm = new SRTM(SRTM::SRTMModel::OneArcSecond);`

**Comment by ABelliqueux** [comment]
Hi, 
I tried the changes you suggest but the results are the same. I guess the SRTM data for this particular region is not good... Not sure what to do next. Maybe I'll contact the webmaster at gpsvisualizer.com about the data he uses...

**Comment by BourgeoisLab** [comment]
Can you post or send me a gpx file with a track in the region.

**Comment by ABelliqueux** [comment]
There you go !

You'll find enclosed in the attachment two gpx files; the 'raw' file as exported from my watch, 
and a version with the elevation data I got after uploading this very same file on https://www.gpsvisualizer.com/elevation , which seems more close to reality. 

[sample-gpx.zip](https://github.com/BourgeoisLab/GPXLab/files/3617036/sample-gpx.zip)

I am also attaching the HGT file you'll need in conjonction with the N47W003.SRTMGL1.hgt.zip file I already attached in a previous message :

[N47W002.SRTMGL1.hgt.zip](https://github.com/BourgeoisLab/GPXLab/files/3617054/N47W002.SRTMGL1.hgt.zip)

**Comment by BourgeoisLab** [comment]
Thanks. I got it.
In strm.cpp change line 155 ans 156 to:
```
bool SRTM::getAltitude(double latitude, double longitude, short &altitude)
{
    int _iLatitude  = (int)floor(latitude);
    int _iLongitude = (int)floor(longitude);
    ....
}
```
I will make the changes in the code later this or next week. Moving to SRTM1 and also changing the averaging function.

**Comment by BourgeoisLab** [comment]
Fixed with commit dc1b17b.
